tests/sources/test_poetry_pyproject.py

```diff
@@ -0,0 +1,289 @@
+    assert capsys.readouterr().err == ''
+
+
+def test_get_python_requires_not_a_string(tmp_path, capsys):
+    pyproject_toml = tmp_path / "pyproject.toml"
+    pyproject_toml.write_text(textwrap.dedent("""\
+        [tool.poetry]
+            name='foo',
+            [tool.poetry.dependencies]
+                python = [">=3.6"]
+    """))
+    assert get_python_requires(str(pyproject_toml)) is None
+    assert (
+        'The value passed to python dependency is not a string'
+        in capsys.readouterr().err
+    )
+
+
+def test_update_python_requires(tmp_path, fix_max_python_3_version):
+    fix_max_python_3_version(7)
+    filename = tmp_path / "pyproject.toml"
+    filename.write_text(textwrap.dedent("""\
+        [tool.poetry]
+            name='foo',
+            [tool.poetry.dependencies]
+                python = ">=3.4"
+    """))
+    result = update_python_requires(str(filename), v(['3.5', '3.6', '3.7']))
+    assert result is not None
+    assert "".join(result) == textwrap.dedent("""\
+        [tool.poetry]
+            name='foo',
+            [tool.poetry.dependencies]
+                python = ">=3.5"
+    """)
+
+
+def test_update_python_requires_file_object(fix_max_python_3_version):
+    fix_max_python_3_version(7)
+    fp = StringIO(textwrap.dedent("""\
+        [tool.poetry]
+            name='foo',
+            [tool.poetry.dependencies]
+                python = ">=3.4"
+    """))
+    fp.name = "pyproject.toml"
+    result = update_python_requires(fp, v(['3.5', '3.6', '3.7']))
+    assert result is not None
+    assert "".join(result) == textwrap.dedent("""\
+        [tool.poetry]
+            name='foo',
+            [tool.poetry.dependencies]
+                python = ">=3.5"
+    """)
+
+
+def test_update_python_requires_when_missing(capsys):
+    fp = StringIO(textwrap.dedent("""\
+        [tool.poetry]
+            name='foo',
+    """))
+    fp.name = "pyproject.toml"
+    result = update_python_requires(fp, v(['3.5', '3.6', '3.7']))
+    assert result is None
+    assert capsys.readouterr().err == ""
+
+
+def test_update_python_requires_preserves_style(fix_max_python_3_version):
+    fix_max_python_3_version(2)
+    fp = StringIO(textwrap.dedent("""\
+        [tool.poetry]
+            name='foo',
+            [tool.poetry.dependencies]
+                python = ">=2.7,!=3.0.*"
+    """))
+    fp.name = "pyproject.toml"
+    result = update_python_requires(fp, v(['2.7', '3.2']))
+    assert "".join(result) == textwrap.dedent("""\
+        [tool.poetry]
+            name='foo',
+            [tool.poetry.dependencies]
+                python = ">=2.7,!=3.0.*,!=3.1.*"
+    """)
+
+
+def test_update_python_requires_multiline(fix_max_python_3_version):
+    fix_max_python_3_version(2)
+    fp = StringIO(textwrap.dedent("""\
+        [tool.poetry]
+            name='foo',
+            [tool.poetry.dependencies]
+                python = ', '.join([
+                '>=2.7',
+                '!=3.0.*',
```

Comment:
Is this actually valid TOML syntax, supported by Poetry?

Reply:
[it should](https://python-poetry.org/docs/dependency-specification/)